sparse/impl/KokkosSparse_gmres_numeric_impl.hpp

```diff
@@ -0,0 +1,391 @@
+          }
+          auto Hj_h = Kokkos::subview(H_h, Kokkos::make_pair(0, j + 1), j);
+        } else if (ortho == GmresHandle::Ortho::CGS2) {
+          auto V0j =
+              Kokkos::subview(V, Kokkos::ALL, Kokkos::make_pair(0, j + 1));
+          auto Hj   = Kokkos::subview(H, Kokkos::make_pair(0, j + 1), j);
+          auto Hj_h = Kokkos::subview(H_h, Kokkos::make_pair(0, j + 1), j);
+          KokkosBlas::gemv("C", one, V0j, Wj, zero, Hj);  // Hj = Vj^T * wj
+          KokkosBlas::gemv("N", -one, V0j, Hj, one, Wj);  // wj = wj - Vj * Hj
+
+          // Re-orthog CGS:
+          auto orthoTmpSub =
+              Kokkos::subview(orthoTmp, Kokkos::make_pair(0, j + 1));
+          KokkosBlas::gemv("C", one, V0j, Wj, zero,
+                           orthoTmpSub);  // tmp (Hj) = Vj^T * wj
+          KokkosBlas::gemv("N", -one, V0j, orthoTmpSub, one,
+                           Wj);                    // wj = wj - Vj * tmp
+          KokkosBlas::axpy(one, orthoTmpSub, Hj);  // Hj = Hj + tmp
+          Kokkos::deep_copy(Hj_h, Hj);
+        } else {
+          throw std::invalid_argument(
+              "Invalid argument for 'ortho'.  Please use 'CGS2' or 'MGS'.");
+        }
+
+        MT tmpNrm     = KokkosBlas::nrm2(Wj);
+        H_h(j + 1, j) = tmpNrm;
+        if (tmpNrm > 1e-14) {
+          Vj = Kokkos::subview(V, Kokkos::ALL, j + 1);
+          KokkosBlas::scal(Vj, one / H_h(j + 1, j), Wj);  // Vj = Wj/H(j+1,j)
+        }
+        Kokkos::Profiling::popRegion();
+
+        // Givens for real and complex (See Alg 3 in "On computing Givens
+        // rotations reliably and efficiently" by Demmel, et. al. 2001) Apply
+        // Givens rotation and compute shortcut residual:
+        for (int i = 0; i < j; i++) {
+          ST tempVal    = CosVal_h(i) * H_h(i, j) + SinVal_h(i) * H_h(i + 1, j);
+          H_h(i + 1, j) = -karith::conj(SinVal_h(i)) * H_h(i, j) +
+                          CosVal_h(i) * H_h(i + 1, j);
+          H_h(i, j) = tempVal;
+        }
+        ST f  = H_h(j, j);
+        ST g  = H_h(j + 1, j);
+        MT f2 = karith::real(f) * karith::real(f) +
+                karith::imag(f) * karith::imag(f);
+        MT g2 = karith::real(g) * karith::real(g) +
+                karith::imag(g) * karith::imag(g);
+        ST fg2        = f2 + g2;
+        ST D1         = one / karith::sqrt(f2 * fg2);
+        CosVal_h(j)   = f2 * D1;
+        fg2           = fg2 * D1;
+        H_h(j, j)     = f * fg2;
+        SinVal_h(j)   = f * D1 * karith::conj(g);
+        H_h(j + 1, j) = zero;
+
+        GVec_h(j + 1) = GVec_h(j) * (-karith::conj(SinVal_h(j)));
+        GVec_h(j)     = GVec_h(j) * CosVal_h(j);
+        shortRelRes   = karith::abs(GVec_h(j + 1)) / nrmB;
+
+        if (verbose) {
+          std::cout << "Shortcut relative residual for iteration "
+                    << j + (cycle * m) << " is: " << shortRelRes << std::endl;
+        }
+        if (tmpNrm <= 1e-14 && shortRelRes >= tol) {
```

Comment:
This 1e-14  (here and in line 177 at the start of the while loop) should really be adjusted based on the scalar type, but I'm not sure off the top of my head the best way to do that. Basically, we are asking whether the tmpNrm is close enough to machine epsilon that we should consider it zero.

Reply:
We can use a meta function like I did for the tol in the new gmres test.